### PR TITLE
quiche: retain filter chain via ProofSource::Details

### PR DESCRIPTION
**Pull request by danzh2010**
Signed-off-by: Dan Zhang <danzh@google.com>

Implement quic::ProofSource::Details to retain filter chain and pass it around during CHLO processing.  Replace the current way of filter chain retrieval upon receiving first packet on a new connection with this new approach.

Use the filter chain to initialize L4 filters when encryption level become forward secure. 

Risk Level: low, not in use
Testing: re-use exiting proof source test. Simplify active_quic_listener_test.cc and envoy_quic_dispatcher_test.cc because of the behavior change.
Part of: #2557

### CODE REVIEW

**Comment by alyssawilk** [comment]
Will let @wu-bin do first pass
@wrowe for windows failure - I'm seeing a bunch of those today.  @danzh2010 don't worry about those as they're not merge blocking yet

File: source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc [hunk 1]
```diff
@@ -39,8 +34,18 @@ std::unique_ptr<quic::QuicCryptoServerStreamBase>
...
+    return std::make_unique<EnvoyQuicCryptoServerStream>(crypto_config, compressed_certs_cache,
+                                                         this, stream_helper());
+  case quic::PROTOCOL_TLS1_3:
+    return std::make_unique<EnvoyQuicTlsServerHandshaker>(this, *crypto_config);
+  case quic::PROTOCOL_UNSUPPORTED:
+    break;
+  }
+  QUIC_BUG << "Unknown handshake protocol: "
+           << static_cast<int>(connection()->version().handshake_protocol);
+  return nullptr;
```
Comment (wu-bin) [inline source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc:48]: I think return a nullptr may crash the program, how about just return a EnvoyQuicCryptoServerStream and let the handshake fail?

File: source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc [hunk 2]
```diff
@@ -103,17 +108,49 @@ void EnvoyQuicServerSession::OnCanWrite() {
...
+  const DetailsWithFilterChain* details =
+      dynamic_cast<const EnvoyQuicTlsServerHandshaker*>(GetCryptoStream())->proofSourceDetails();
+  if (details == nullptr) {
+    ENVOY_CONN_LOG(
+        trace,
+        "ProofSource didn't provide ProofSource::Details. No filter chain will be installed.",
+        *this);
+    return;
+  }
+  createNetworkFilters(details->filterChain());
```
Comment (wu-bin) [inline source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc:146]: Can we move this into a function and call it from SetDefaultEncryptionLevel and OnOneRttKeysAvailable?

File: source/extensions/quic_listeners/quiche/envoy_quic_crypto_server_stream.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      std::unique_ptr<quic::ProcessClientHelloResultCallback> done_cb) override {
+    auto done_cb_wrapper =
+        std::make_unique<EnvoyProcessClientHelloResultCallback>(this, std::move(done_cb));
+    ASSERT(done_cb_wrapper_ == nullptr);
+    done_cb_wrapper_ = done_cb_wrapper.get();
+    auto details = static_cast<DetailsWithFilterChain*>(proof_source_details.get());
+    if (details != nullptr) {
+      // Retain a copy of the old proof source details.
+      details_ = std::make_unique<DetailsWithFilterChain>(*details);
+    }
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/envoy_quic_crypto_server_stream.h]: It shouldn't be retained here. I was following the internal implementation which letch in both places. But due to later changes, GetProof() should never be called before ProcessClientHello() and thus the proof_source_details should always be nullptr.

**Comment by danzh2010** [comment]
Ping @alyssawilk

File: source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc [hunk 4]
```diff
@@ -103,17 +106,39 @@ void EnvoyQuicServerSession::OnCanWrite() {
...
 void EnvoyQuicServerSession::OnOneRttKeysAvailable() {
   quic::QuicServerSessionBase::OnOneRttKeysAvailable();
   raiseConnectionEvent(Network::ConnectionEvent::Connected);
+  maybeCreateNetworkFilters();
+}
+
+void EnvoyQuicServerSession::maybeCreateNetworkFilters() {
+  const EnvoyQuicProofSourceDetails* proof_source_details =
+      dynamic_cast<const EnvoyCryptoServerStream*>(GetCryptoStream())->proofSourceDetails();
+  if (proof_source_details == nullptr) {
```
Comment (alyssawilk) [inline source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc:128]: Is this something worth fixing?  We do try to keep the unit tests accurate, even if we do something like mock out proofSourceDetails to return a mock pointer

File: source/extensions/quic_listeners/quiche/envoy_quic_crypto_server_stream.cc [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+void EnvoyQuicCryptoServerStream::ProcessClientHello(
+    quic::QuicReferenceCountedPointer<quic::ValidateClientHelloResultCallback::Result> result,
+    std::unique_ptr<quic::ProofSource::Details> proof_source_details,
+    std::unique_ptr<quic::ProcessClientHelloResultCallback> done_cb) {
+  auto done_cb_wrapper =
+      std::make_unique<EnvoyProcessClientHelloResultCallback>(this, std::move(done_cb));
+  ASSERT(done_cb_wrapper_ == nullptr);
+  done_cb_wrapper_ = done_cb_wrapper.get();
+  ASSERT(proof_source_details == nullptr);
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/envoy_quic_crypto_server_stream.cc:42]: done

File: source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc (hunk 4 above)
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc]: Makes sense, I added a MockFilterChain in TestProofSource. And ASSERT proof_source_details != nullptr here.

**Comment by danzh2010** [comment]
Fixed GCC failure. PTAL!

**Pull request merged**